### PR TITLE
Add Sigmoid Cross Entropy Loss

### PR DESCRIPTION
**Pull request by tanmayb123**
a.k.a. "binary_crossentropy" from Keras

### CODE REVIEW

**Comment by tanmayb123** [comment]
I signed it!

File: Sources/DeepLearning/Loss.swift [hunk 1]
```diff
@@ -27,3 +27,12 @@ public func softmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
     logits: Tensor<Scalar>, labels: Tensor<Scalar>) -> Tensor<Scalar> {
     return -(labels * logSoftmax(logits)).mean(alongAxes: 0).sum()
 }
+
```
Comment (saeta) [inline Sources/DeepLearning/Loss.swift]: We haven't been doing this well yet, but could you add a Swift documentation comment (`///`-prefixed) describing this function? Maybe something along the lines of:

```
/// Computes the sigmoid cross entropy between logits and labels.
///
/// Logits are one-hot encoded outputs from a neural network. Labels are integer values that correspond to the correct output.
```

or something along those lines?

Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: Alright, will do.

**Comment by tanmayb123** [comment]
Thanks @saeta :)

I've added documentation comments for each loss function.

File: Sources/DeepLearning/Loss.swift [hunk 2]
```diff
@@ -16,14 +16,38 @@
 import TensorFlow
 #endif
 
+/// Computes the Mean Squared Error loss between logits and labels
+///
+/// - Parameters
+///     - logits: one-hot encoded outputs from a neural network.
```
Comment (dan-zheng) [inline Sources/DeepLearning/Loss.swift]: Please indent individual parameter doc comments by 2, to match the existing style.
```diff
 /// - Parameters
-///     - logits: one-hot encoded outputs from a neural network.
+///   - logits: one-hot encoded outputs from a neural network.
```
Thank you!

Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: Done

File: Sources/DeepLearning/Loss.swift [hunk 3]
```diff
@@ -16,14 +16,38 @@
 import TensorFlow
 #endif
 
+/// Computes the Mean Squared Error loss between logits and labels
```
Comment (dan-zheng) [inline Sources/DeepLearning/Loss.swift]: Please end each doc comment with a period.
```suggestion
/// Computes the mean squared error between logits and labels.
```

Also, I think you can drop the word "loss".
Example from [TensorFlow Python docs](https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits_v2):
>  Computes softmax cross entropy between logits and labels.

Thank you!

Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: Done

File: Sources/DeepLearning/Loss.swift [hunk 4]
```diff
@@ -16,14 +16,38 @@
 import TensorFlow
 #endif
 
+/// Computes the mean squared error between logits and labels.
+///
+/// - Parameters
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
/// - Parameters:
```

File: Sources/DeepLearning/Loss.swift [hunk 5]
```diff
@@ -16,14 +16,38 @@
...
+///   - labels: one-hot encoded values that correspond to the correct output.
 @differentiable
 public func meanSquaredError<Scalar: TensorFlowFloatingPoint>(
     predicted: Tensor<Scalar>, expected: Tensor<Scalar>) -> Tensor<Scalar> {
     return (expected - predicted).squared().mean()
 }
 
+/// Computes the softmax cross entropy (categorical cross entropy) between logits and labels.
+///
+/// - Parameters
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
/// - Parameters:
```

File: Sources/DeepLearning/Loss.swift [hunk 6]
```diff
@@ -16,14 +16,38 @@
...
+///   - labels: one-hot encoded values that correspond to the correct output.
 @differentiable
 public func softmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
     logits: Tensor<Scalar>, labels: Tensor<Scalar>) -> Tensor<Scalar> {
     return -(labels * logSoftmax(logits)).mean(alongAxes: 0).sum()
 }
+
+/// Computes the sigmoid cross entropy (binary cross entropy) between logits and labels.
+///
+/// - Parameters
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
/// - Parameters:
```

File: Sources/DeepLearning/Loss.swift [hunk 7]
```diff
@@ -16,14 +16,38 @@
...
+///   - logits: single continuous values from 0 to 1.
+///   - labels: integer values that correspond to the correct output.
+@differentiable
+public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
+    logits: Tensor<Scalar>, labels: Tensor<Scalar>) -> Tensor<Scalar> {
+    let loss = labels * log(logits) +
+               (Tensor<Scalar>(ones: labels.shape) - labels) *
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
         (Tensor<Scalar>(ones: labels.shape) - labels) *
```

File: Sources/DeepLearning/Loss.swift [hunk 8]
```diff
@@ -16,14 +16,38 @@
...
+               log(Tensor<Scalar>(ones: logits.shape) - logits)
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
        log(Tensor<Scalar>(ones: logits.shape) - logits)
```

File: Sources/DeepLearning/Loss.swift [hunk 9]
```diff
@@ -16,14 +16,38 @@
...
 @differentiable
 public func softmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
     logits: Tensor<Scalar>, labels: Tensor<Scalar>) -> Tensor<Scalar> {
     return -(labels * logSoftmax(logits)).mean(alongAxes: 0).sum()
 }
+
+/// Computes the sigmoid cross entropy (binary cross entropy) between logits and labels.
+///
+/// - Parameters
+///   - logits: single continuous values from 0 to 1.
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]: Start every parameter description with an uppercase letter.

File: Sources/DeepLearning/Loss.swift [hunk 10]
```diff
@@ -16,14 +16,38 @@
...
+///   - labels: integer values that correspond to the correct output.
+@differentiable
+public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
+    logits: Tensor<Scalar>, labels: Tensor<Scalar>) -> Tensor<Scalar> {
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]: Move `) -> Tensor<Scalar> {` to the next line.

See line wrapping guidelines here: https://google.github.io/swift/#line-wrapping.

File: Sources/DeepLearning/Loss.swift (hunk 9 above)
Comment (rxwei) [reply Sources/DeepLearning/Loss.swift]: And wrap backticks around numeric values (`0`, `1`).

File: Sources/DeepLearning/Loss.swift (hunk 7 above)
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]: The call to `init(ones:)` is unnecessary and inefficient. Since `-` is broadcasting, yon can just use `1 -`. The whole statement can be simplified to:
```swift
let loss = labels * log(logits) + (1 - labels) * log(1 - logits)
```

Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: I did try that at first, but got this error:

```
error: repl.swift:23:2: error: function is not differentiable
@differentiable
~^~~~~~~~~~~~~~

repl.swift:24:13: note: when differentiating this function definition
public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
            ^

repl.swift:26:42: note: function call is not differentiable because generic requirements are not met
    let loss = labels * log(logits) + (1 - labels) * log(1 - logits)
                                         ^


```

Comment (rxwei) [reply Sources/DeepLearning/Loss.swift]: Oh, in that case it's definitely a high-priority bug.  Could you file a bug on bugs.swift.org for this as well?

Comment (rxwei) [reply Sources/DeepLearning/Loss.swift]: Could you also try:
```swift
let loss = labels * log(logits) + (Scalar(1) - labels) * log(Scalar(1) - logits)
```
?

File: Sources/DeepLearning/Loss.swift (hunk 10 above)
Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: Done - thanks!

File: Sources/DeepLearning/Loss.swift (hunk 7 above)
Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: Same error (had to split it up into chunks because `the compiler is unable to type-check this expression in reasonable time`):

```
error: repl.swift:23:2: error: function is not differentiable
@differentiable
~^~~~~~~~~~~~~~

repl.swift:24:13: note: when differentiating this function definition
public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
            ^

repl.swift:28:29: note: function call is not differentiable because generic requirements are not met
        let p2 = (Scalar(1) - labels)
                            ^


```

Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: Filed the bug report:

File: Sources/DeepLearning/Loss.swift [hunk 11]
```diff
@@ -16,14 +16,39 @@
...
+
+/// Computes the sigmoid cross entropy (binary cross entropy) between logits and labels.
+///
+/// - Parameters:
+///   - logits: Single continuous values from `0` to `1`.
+///   - labels: Integer values that correspond to the correct output.
+@differentiable
+public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
+    logits: Tensor<Scalar>, labels: Tensor<Scalar>
+    ) -> Tensor<Scalar> {
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
) -> Tensor<Scalar> {
```

File: Sources/DeepLearning/Loss.swift (hunk 7 above)
Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: This does work:

```
@differentiable
public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
    logits: Tensor<Scalar>, labels: Tensor<Scalar>
) -> Tensor<Scalar> {
    let loss = labels * log(logits) +
        (Tensor<Scalar>(1) - labels) *
        log(Tensor<Scalar>(1) - logits)
    return -loss.mean(alongAxes: 0).sum()
}
```

Comment (rxwei) [reply Sources/DeepLearning/Loss.swift]: This looks good, please use that for now. You can move `log(Tensor<Scalar>(1) - logits)` to the end of the second line of the statement. We are enforcing a horizontal limit of 100 columns.

Comment (tanmayb123) [reply Sources/DeepLearning/Loss.swift]: Done. Thanks!

**Comment by rxwei** [comment]
Thanks Tanmay!
